### ISSUE TITLE
Automatic interception for iOS

### ISSUE THREAD

**Issue by pimterry** [httptoolkit/feedback#11]

**Comment by nutterthanos** [comment]
yeah i don't think this would be possible due to Apple's restrictions on IOS unless people jailbreak

**Comment by SeriousM** [comment]
any updates on ios interception?

**Comment by pimterry** [comment]
No major updates on automating this yet, but it is completely possible to manually intercept iOS devices already. There's a new guide available for the manual setup steps to do so here: https://httptoolkit.com/docs/guides/ios/

**Comment by shirshak55** [comment]
@pimterry 
I think we need HTTP toolkit app similar to the android one that sets VPN so we can confirm the proxy is working properly. I have noticed few apps can bypass proxy if they want.

**Comment by pimterry** [comment]
Hmmm, ok. That's awkward but yes it certainly sounds possible.

Any idea how they're doing that? If you have any examples of apps that do this then that would be very helpful.

**Comment by shirshak55** [comment]
@pimterry I think there is a way to bypass proxy from the app side and many apps do it. 

https://stackoverflow.com/questions/28061353/how-to-bypass-proxy-in-ios-programmatically

The only problem is the app doesn't tell us it is bypassing the proxy. However, I am sure they can't bypass VPN. So, I was thinking there should be an IOS app that will create VPN and connect to HTTP toolkit app.

**Comment by KevinRodriguez777** [comment]
are there any planned updates to this?

**Comment by pimterry** [comment]
This is now being actively worked on :smile:. Hoping for a fully automated solution later this year or very early 2024. In the meantime, it's still quite possible to manually intercept iOS, you just need to follow the guide here: https://httptoolkit.com/docs/guides/ios/

If you have any improvements to that process, PRs are welcome, the markdown for that doc is here: https://github.com/httptoolkit/httptoolkit-website/blob/main/src/docs/guides/ios.md

**Comment by moazam1** [comment]
First of all thank you for your hard work. I am not sure if it's the right place to post an issue about ios interception. I setup using above guidelines and it works great for the post of the apps but few apps just does not recognize the network connections example: Tuya and Smart Life. I also noticed few entries like:

Aborted connection to p164-contacts.icloud.com

**Comment by pimterry** [comment]
I have now have a working prototype for the first iteration of automated iOS support!

In this initial first release version, this requires a jailbroken device with Frida running connected to your computer.

Once you have that in place, it's possible to intercept any single target app totally automatically, without needing to capture the entire system - including disabling the vast majority of certificate pinning mechanisms as well.

This jailbreak+Frida setup will be relatively common among existing reverse engineers, and I'll put together some docs later to help with that setup, but I am also intending to eventually widen this further later for other setups. As discussed above, it should also be possible to provide this exact same interception for any applications running in iOS simulators (on Mac only) and for _debuggable_ apps running on real devices. Hopefully that will cover most use cases.

Intercepting non-debuggable apps on non-jailbroken devices will never be possible totally automatically (AFAICT) but the manual process is reasonably quick & effective already, and I'm planning to tighten that up further to improve the UX there soon.

**I'm looking for testers for this**. The prototype scripts will be available for real-world alpha testing sometime in the next week or two. If you have a jailbroken device and you'd like to try this out, please send me an email at tim@httptoolkit.com.

**Comment by pimterry** [comment]
For jailbroken devices, this is now live: https://x.com/HttpToolkit/status/1803031754363543986 :rocket:

![image](https://github.com/httptoolkit/httptoolkit/assets/1526883/d43f9fb5-ed3f-4602-bd8f-d4756d23395a)

If you have a jailbroken iPhone with Frida, you can click a button to see every app on the phone, pick any app, and it'll be automatically intercepted (in isolation with no other system traffic) and all traffic will be captured directly (with certificate pinning disabled automatically). No manual HTTP Toolkit setup required whatsoever.

I'll keep this issue open as we expand this further, aiming for:

* Blog posts explaining how this works, and detailed walkthroughs to get started
* Semi-automated setup for non-Frida interception
* Automated setup for iOS simulator interception on Mac
* Guide (and maybe automation) for non-jailbroken Frida interception of debuggable apps, via [Frida Gadget](https://frida.re/docs/gadget/)

Note that manual whole-device iOS interception is also fully supported, with just a couple of steps required and now with a QR code for easier setup - see https://httptoolkit.com/docs/guides/ios/ for more info.

**Comment by secureITmania** [comment]
We need a feature to allow the addition of an external Frida script in order to bypass jailbreak detection and run the application smoothly.

**Comment by secureITmania** [comment]
> @secureITmania that's a very broad topic. I'm open to inclusions like that, but it requires some separate investigation and probably manual testing. If you find challenging cases where that's required, I'd suggest you use the Frida scripts directly so you have full control (guide here: https://github.com/httptoolkit/frida-interception-and-unpinning/?tab=readme-ov-file#ios-getting-started-guide), test out different jailbreak detection fixes using Frida directly, and if you find any good approaches you open an issue on that Frida scripts repo and those can be included there directly.

Yes, I have a custom Frida script to bypass the jailbreak detection. It is working fine. However, if I run the "iOS App via Frida," it will relaunch the app and display a jailbreak detection banner.
